backend/src/test/controller/PromotionController.test.ts

```diff
@@ -0,0 +1,339 @@
+import { UserFactory } from '../factory/UserFactory';
+import { registerTestApplication } from './BaseController';
+import { PromotionFactory } from '../factory/PromotionFactory';
+import { DiscountFactory } from '../factory/DiscountFactory';
+import { ScheduleFactory } from '../factory/ScheduleFactory';
+import { PromotionRepository } from '../../main/repository/PromotionRepository';
+import { DiscountType } from '../../main/data/DiscountType';
+import { Promotion } from '../../main/entity/Promotion';
+
+describe('Unit tests for PromotionController', function () {
+  let userRepository: UserRepository;
+  let promotionRepository: PromotionRepository;
+  let app: Express;
+
+  beforeAll(async () => {
+    await connection.create();
+    app = registerTestApplication();
+  });
+
+  afterAll(async () => {
+    await connection.close();
+  });
+
+  beforeEach(async () => {
+    await connection.clear();
+    userRepository = getCustomRepository(UserRepository);
+    promotionRepository = getCustomRepository(PromotionRepository);
+  });
+
+  test('GET /promotions', async (done) => {
+    const user: User = new UserFactory().generate();
+    const discount = new DiscountFactory().generate();
+    const schedule = new ScheduleFactory().generate();
+    const promotion = new PromotionFactory().generate(user, discount, [
+      schedule,
+    ]);
+
+    await userRepository.save(user);
+    await promotionRepository.save(promotion);
+
+    request(app)
+      .get('/promotions')
+      .expect(200)
+      .end((err, res) => {
+        if (err) return done(err);
+        const promotions = res.body;
+        expect(promotions).toHaveLength(1);
+        comparePromotions(promotions[0], promotion);
+        done();
+      });
+  });
+
+  test('GET /promotions - query parameters without search query', async (done) => {
+    const user: User = new UserFactory().generate();
+
+    const promotion1 = new PromotionFactory().generate(
+      user,
+      new DiscountFactory().generate(DiscountType.PERCENTAGE),
+      [new ScheduleFactory().generate()]
+    );
+    const promotion2 = new PromotionFactory().generate(
+      user,
+      new DiscountFactory().generate(DiscountType.AMOUNT),
+      [new ScheduleFactory().generate()]
+    );
+    const promotion3 = new PromotionFactory().generate(
+      user,
+      new DiscountFactory().generate(DiscountType.OTHER),
+      [new ScheduleFactory().generate()]
+    );
+
+    await userRepository.save(user);
+    await promotionRepository.save(promotion1);
+    await promotionRepository.save(promotion2);
+    await promotionRepository.save(promotion3);
+
+    request(app)
+      .get('/promotions')
+      .query({
+        discountType: DiscountType.PERCENTAGE,
+      })
+      .expect(200)
+      .end((err, res) => {
+        if (err) return done(err);
+        const promotions = res.body;
+        expect(promotions).toHaveLength(1);
+        comparePromotions(promotions[0], promotion1);
+        done();
```

Comment:
Also just wondering, is the difference in this format vs. awaiting the call to `request(app)...expect(200)` just a stylistic choice?

Reply:
Ohh yeah, actually that's a good point I didn't realize thanks for bringing it up.
I noticed I kind of mix. 
```
await request(app).get(...)
// do something after

vs

request(app).get(...)
   .end((err, res) => {
       // do something after
   })
```

I can make this more consistent actually. I'm kind of thinking of removing all the repository calls in the `// do something after` sections because of the problem mentioned here https://github.com/ubclaunchpad/foodies/pull/94#issuecomment-752329183 (not the deadlock issue, but how the transaction hasn't committed and how our repository calls would not get up to date information).